### ISSUE TITLE
Сделать нормальный биндинг методов

### ISSUE THREAD

**Issue by Znack** [Burize/FULLSTACK-GameOfLife#8]
В контроллере можно уйти от `() => this.startGame.call(this)`.
Для этого надо изучить разные способы биндинга методов в классе. опиши их сюда и напиши, какой хотел бы использовать, я помогу с выбором

**Comment by Znack** [comment]
ну и `const _this = this;` в 99% случаев не нужен уже, есть же как минимум стрелочные функции, изучи как там контекст работает (и как там arguments работают, очень полезно будет, поможет кучи граблей на твой лоб не просыпать)

**Comment by Burize** [comment]
Я использовал(как я понял) самый очевидный:  this.startGame = this.startGame.bind(this);

В ES7 появился новый синтаксис для этого:  this.startGame = ::this.startGame; (но у меня он не работает, наверное устарел babel-loader)

Есть такая штука https://github.com/sindresorhus/auto-bind . Получается удобно, не нужно каждый метод биндить, но тут нужно скачивать внешний модуль(библиотеку).

Есть пара примеров с React, но наверно пока рано.

Много, где есть примеры с arrow function, но они тоже не работают у меня х_х
https://forums.meteor.com/t/no-more-need-for-method-pre-binding-looks-like/26866/5  (хотя тут тоже с реактом)


"Для этого надо изучить разные способы биндинга методов в классе." - как я понял подразумевается, что их как минимум несколько должно быть. Самый распространенный и очевидный - тот, который я использовал. Воот .... а других я не нашел. Почти все примеры или с реактом или не с классами.

**Comment by Znack** [comment]
На самом деле смотри примеры и с  React в плане биндинга там обычные js конструкции используют, ничего чисто реактовского.
Первые два абзаца не самые лучшие варианты, потому что биндить ты будешь в момент использования метода, а не в момент создания. И если у тебя метод дофига популярный, то ты будешь десятки раз этот бинд на метод вызывать. плюс если метод публичный и им будут пользоваться извне, внешний код должен знать про биндинг и использовать его, что не самый лучший пример API класса :)

А ты понимаешь как autobind работает в целом? Если да, то можешь сразу еще более ручной и простой способ предложить (я так понял помимо первых двух, которые как я сказал, не лучший вариант, и вот autobind ты предложил только стрелочные функции, но  autobind точно не через стрелочные функции эту проблему решает)

**Comment by Burize** [comment]
там в прототип записывается, а не в свойство класса. 
this.startGame = this.startGame.bind(this); - так получается совсем не правильно. Из-за этого были ошибки в тестах:
  spy_changeHeight = sinon.spy(game, 'changeHeight') - следил за методом класса, а при игре вызывались переменные.

Как я понял надо  this.\_\_proto__.startGame = this.startGame.bind(this);
Только вот так тесты совсем не работают.

Например тест (самый последний в контроллере) на изменение ширины - там ошибки в том, что не вызывается функция changeWidth у контроллера и у модели, и сама ширина тоже не меняется. Я понял бы, если бы не вызвалась функция, но ширина изменилась бы. Тогда была бы проблема в spy.

Как добавление __proto__ отразилось на том, что перестали работать(вызываться) функции Модели? - Сама игра работает, падают только тесты.


Причем в хроме (всмысле тестировании в хроме) действительно работает неправильно ... но а просто если запустить то все ок

**Comment by Znack** [comment]
```
this.__proto__.startGame = this.startGame.bind(this);
``` 
- даже больше, если ты так сделаешь, ты привяжешь метод класса к одному инстансу, и если второй инстанс метод вызовет, this будет у первого 😆 Опааасно...
А с spy_changeHeight = sinon.spy(game, 'changeHeight') что именно не так? Здесь вроде должно работать на самом деле. game — это же инстанс класса? Синон вообще не должен лезти в прототип, он должен у инстанса именно метод подменить. Надо сюда копать, потому что тут все должно рабоатть

**Comment by Burize** [comment]
> А с spy_changeHeight = sinon.spy(game, 'changeHeight') что именно не так? Здесь вроде должно работать на самом деле. game — это же инстанс класса?

Как я понял он следит за вызовом метода класса: spy_changeHeight = sinon.spy(game, 'changeHeight'), а при игре вызывается не метод класса, а обертка: this.startGame = this.startGame.bind(this);

**Comment by Burize** [comment]
> даже больше, если ты так сделаешь, ты привяжешь метод класса к одному инстансу, и если второй инстанс метод вызовет, this будет у первого 😆 Опааасно...
 
ну да, прототип один у всех экземпляров. Получается у всех будет контекст последнего созданного

**Comment by Burize** [comment]
> А ты понимаешь как autobind работает в целом? Если да, то можешь сразу еще более ручной и простой способ предложить


    module.exports = self => {
	for (const key of Object.getOwnPropertyNames(self.constructor.prototype)) {  // для всех полей(ключей) в прототипе (собственных)
		const val = self[key]; //получает это поле(свойство)

		if (key !== 'constructor' && typeof val === 'function') { // проверяет, чтобы это было методом
			self[key] = val.bind(self); // присваивает ему обертку (биндит)
		}
	}
	return self;
         };


получается это тоже самое, что и this.startGame = this.startGame.bind(this); ?

**Comment by Znack** [comment]
да-да, это то же самое, что и бинд, но есть важное отличие. Ты его вызываешь один раз в конструкторе и все :) Это как раз та деталь, которая все решает

> Как я понял он следит за вызовом метода класса: spy_changeHeight = sinon.spy(game, 'changeHeight'), а при игре вызывается не метод класса, а обертка: this.startGame = this.startGame.bind(this);

не, sinon должен уметь и нормальные методы переопределять, котороые были забиндены в конструкторе. Читай примеры, изучай работу с sinon, потом если не получится, пиши подробный отчет что и как изучил, я тебе помогу

**Comment by Burize** [comment]
> Я использовал(как я понял) самый очевидный: this.startGame = this.startGame.bind(this);

> Первые два абзаца не самые лучшие варианты, потому что биндить ты будешь в момент использования метода, а не в момент создания. И если у тебя метод дофига популярный, то ты будешь десятки раз этот бинд на метод вызывать. плюс если метод публичный и им будут пользоваться извне, внешний код должен знать про биндинг и использовать его, что не самый лучший пример API класса :)

> да-да, это то же самое, что и бинд, но есть важное отличие. Ты его вызываешь один раз в конструкторе и все :) Это как раз та деталь, которая все решает

Но так там отличие только в том, что так они все забиндится при вызове auto-bind, а без этого придется к каждому прописывать биндинг

      constructor(){

     this.updateField = this.updateField.bind(this);

    this.startGame = this.startGame.bind(this);

    this.pauseGame = this.pauseGame.bind(this);

    this.changeCell = this.changeCell.bind(this);

    this.changeWidth = this.changeWidth.bind(this);
     }

> И если у тебя метод дофига популярный, то ты будешь десятки раз этот бинд на метод вызывать. плюс если метод публичный и им будут пользоваться извне, внешний код должен знать про биндинг и использовать его, что не самый лучший пример API класса :)

Эти проблемы также останутся.

**Comment by Znack** [comment]
Ну на самом деле мы немалое время так и делали 😆 То есть биндили каждый метод в конструкторе и сейчас где-нибудь такой легаси всплыть может. В принципе это не худшее решение. И тогда можно мокать метод в прототипе (только в прототипе, а не в инстансе) и все будет работать

**Comment by Burize** [comment]
Исправил. 

https://stackoverflow.com/questions/25578112/spying-on-a-method-with-sinon-method-bound-to-event-listener-method-was-execut

https://stackoverflow.com/questions/15568140/testing-callback-called-after-a-event-trigger-with-sinon-js

https://stackoverflow.com/questions/9113186/backbone-js-click-event-spy-is-not-getting-called-using-jasmine-js-and-sinon-js

Там везде (ну на самом деле не так много вопросов/ответов по этой проблеме) пишут, что нужно задать spy для метода прототипа и не инстанса, причем нужно сделать это до объявления инстанса.
Ну, то, что нужно задавать для метода прототипа, а не инстанса, я могу понять следуя ранее написанному

> Как я понял он следит за вызовом метода класса: spy_changeHeight = sinon.spy(game, 'changeHeight'), а при игре вызывается не метод класса, а обертка: this.startGame = this.startGame.bind(this);

Вооот, но какая разница до или после задать spy для метода прототипа (если после объявления инстанса задать, то он не отследить выполнение)?  Ведь при создании экземпляра класса в прототипе ничего не переопределяется и не оборачивается.

**Comment by Znack** [comment]
Ага, по логике должно быть без разницы. Там скорее всего либо что-то узко специализорованное рекомендовали, что к тебе и не относится, либо что-то сами неверно понимали, либо мы что-то не знаем :)
Ну тебе удалось запустить и с моканием метода после создания инстанса?
Если нет, то мы на собесе просто вместе развернем тогда и подебажим просто
